### ISSUE TITLE
Oddly, example-ruby-sinatra runs slow with k8s as scheduler

### ISSUE THREAD

**Issue by krancour** [deis/deis#4263]
I'm having a tough time getting to the bottom of this.  Opening the issue just to ensure it's tracked.

I've discovered, by chance, that when I deploy the example-ruby-sinatra app using k8s as the scheduler, responses from the application are very slow.  Oddly this does not happen with Fleet as the scheduler _and_ it does not happen with other apps.

Digging in farther, if you `nse` to a container running the sinatra app and `curl -L http://localhost:5000`, it responds very quickly.  If however, you make the request using the IP on the overlay network that Docker has assigned to the container _or_ if you use the IP that k8s uses for the service, then the response takes 10+ seconds.

Compare this to when Fleet is used as the scheduler and requests made using the IP on the overlay network that Docker has assigned to the container are very fast.

Something is fishy and I am still on the hunt.

**Comment by krancour** [comment]
With Fleet:

```
core@ip-10-21-1-112 ~ $ nse 499f29b416a6 # the container in question
slug@499f29b416a6:~$ time curl http://localhost:5000
Powered by Deis!
Running on container ID 499f29b416a6

real	0m0.027s
user	0m0.004s
sys	0m0.006s
slug@499f29b416a6:~$ exit
exit

core@ip-10-21-1-112 ~ $ time curl http://localhost:49154 # the host port mapped to container 5000
Powered by Deis!
Running on container ID 499f29b416a6

real	0m0.019s
user	0m0.002s
sys	0m0.003s

core@ip-10-21-1-112 ~ $ time curl http://10.244.9.120:5000 # container's ip on overlay network
Powered by Deis!
Running on container ID 499f29b416a6

real	0m0.018s
user	0m0.004s
sys	0m0.001s
```

^ All good.

With k8s:

```
core@ip-10-21-1-111 ~ $ nse 273f0fdb7841 # the container in question
slug@taller-aqualung-v2-web-7layl:~$ time curl http://localhost:5000
Powered by Deis!
Running on container ID taller-aqualung-v2-web-7layl

real	0m0.047s
user	0m0.003s
sys	0m0.008s
slug@taller-aqualung-v2-web-7layl:~$ exit
exit

core@ip-10-21-1-111 ~ $ time curl http://10.244.7.122:5000 # container's ip on overlay network
Powered by Deis!
Running on container ID taller-aqualung-v2-web-7layl

real	0m10.026s
user	0m0.004s
sys	0m0.002s

core@ip-10-21-1-111 ~ $ time curl http://10.100.80.91     
Powered by Deis!
Running on container ID taller-aqualung-v2-web-7layl

real	0m10.035s
user	0m0.002s
sys	0m0.003s
```

Note that the k8s service seems _not_ to be the problem, since the timing on accessing via the container's individual IP on the overlay network is _also_ bad.

To further establish this, I manually deleted the service and took the timing again just using the container's IP on the overlay network:

```
kubectl delete service app-46245 --namespace=taller-aqualung
```

```
core@ip-10-21-1-111 ~ $ time curl http://10.244.7.122:5000
Powered by Deis!
Running on container ID taller-aqualung-v2-web-7layl

real	0m10.026s
user	0m0.001s
sys	0m0.006s
```

**Comment by bacongobbler** [comment]
I think this is why @aledbf created https://github.com/aledbf/kube-haproxy-router which bypasses kube-proxy entirely. I'd take a look at where the traffic is going and see if you end up with the same findings.

**Comment by krancour** [comment]
I think I've ruled out kube-proxy and flannel as being part of the problem.  If I curl the Sinatra app _from the host that container is on_ (important) using that container's IP on the overlay network, neither kube-proxy nor flannel should be involved.  I can confirm this by looking at the route tables with `route -n` (from the host).  Such a request is routed directly to the `docker0` interface.

Of course, the above is true regardless of which scheduler is user and regardless of which example app is used.  The question is why is the combination of Sinatra and k8s slow, while no other scheduler / example app combination I have tried has that problem...

Unlike Fleet, where we use unit files that use the `docker` CLI, when using k8s, the kubelet interacts with Docker via Docker's API.  That provides more finely-grained control over the options on the containers.  Using `docker inspect`, I can see some of the many differences between a container scheduled via Fleet and one scheduled via k8s.  None jump out at me as relevant, however, I think some of these differences, combined with Sinatra add up to something meaningful.

What I'm doing right now is adding in puma to the Sinatra app since that's a more proper Ruby-based web server.  I'm curious to see if that solves the problem.

**Comment by krancour** [comment]
Using puma fixes it.  There's got do be something about plain Sinatra's build in web server that doesn't jive nicely with the docker container as created by a kubelet.

**Comment by krancour** [comment]
I still can't spot what Puma is doing differently than plain Sinatra.  I even went as far as to customize the builder to install `netstat` and `ifconfig` on all app images to help get more insight.

It would be good to get to the bottom of this issue.  #4222 has us considering a new default scheduler.  It would be good to have a handle on the conditions that would make a trivial app such as this crawl with k8s as the scheduler.

**Comment by jgmize** [comment]
> I still can't spot what Puma is doing differently than plain Sinatra. I even went as far as to customize the builder to install netstat and ifconfig on all app images to help get more insight.

Is Puma bound to a different address than Sinatra? I believe the default for Sinatra is to bind to 0.0.0.0, and if Puma is bound to a specific interface it could potentially be eliminating some local routing issue. Would you mind adding the results of `netstat -tln` within the container and `ip route show` from both the host and within the container to this issue?

**Comment by mboersma** [comment]
I have also seen this behavior in testing, just FYI: example-ruby-sinatra was crazy slow with k8s.

**Comment by jgmize** [comment]
Has any progress been made on identifying the root cause of the slowness of some apps under k8s? As an additional data point, while testing [bedrock](https://github.com/mozilla/bedrock) under both fleet and k8s, we see a significant slowdown under k8s.

**Comment by krancour** [comment]
@jgmize no... this issue hasn't really been a priority, but I am now actually wondering how this works out with _just_ k8s and Deis not in the equation at all.  I will try that when I have a chance.

**Comment by aledbf** [comment]
@krancour this could be a fix for this issue using `k8s >= v1.1.1`
```
for node in `k get nodes -o name`; do kubectl annotate $node "net.beta.kubernetes.io/proxy-mode=iptables"; done
then restart kube-proxy

Independent reports were: if you assume pod-to-pod (without service proxy) is 100%, service proxy was 40-50%, iptables is > 96%

the default won't flip until kube 1.2, in Feb or March
````

https://kubernetes.slack.com/archives/kubernetes-users/p1447695833001646

**Comment by krancour** [comment]
Since the k8s tech preview has now been removed from the 1.x line and I cannot replicate this issue in Deis v2, I'm closing this.
